logger/factory.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package logger
+
+import (
+	"fmt"
+	"github.com/uber-common/bark"
+)
+
+// LoggerFactory wraps a bark.Logger and provides a way to create named loggers
+// restricted to a specific Level.
+type LoggerFactory struct {
+	logger bark.Logger
+	cache  map[string]*namedLogger
+}
+
+func NewLoggerFactory(l bark.Logger) *LoggerFactory {
+	return &LoggerFactory{
+		logger: l,
+		cache:  make(map[string]*namedLogger),
+	}
+}
+
+// SetLevel sets the minimum level for a named logger. A named logger emits
+// messages with a level equal to or greater than this level.
+func (lf *LoggerFactory) SetLevel(name string, level Level) error {
+	if level < lowestLevel {
+		return fmt.Errorf("level must be higher than %s", lowestLevel)
+	}
+	if level > highestLevel {
+		return fmt.Errorf("level must be lower than %s", highestLevel)
+	}
+	if named, ok := lf.cache[name]; ok {
+		named.setLevel(level)
+	} else {
+		lf.cache[name] = &namedLogger{
+			logger: lf.logger,
+			min:    level,
+		}
+	}
+	return nil
+}
```

Comment:
Setting the level on an individual logger kind of defeats the purpose of having the factory. The levels is what we want to abstract by removing the `name: Level` mapping from user's sight. Another thing to keep in mind is that all loggers created by calling `WithField`/`WithFields` need to preserve the notion of name and level restriction, otherwise we won't be able to change the configuration for all loggers belonging to the same name.

Consider this (ringpop) usecase:
```go
named := factory.Logger("named")
named.WithField("a", 1).Debug("msg")
named.WithField("b", 2).Debug("msg")
```

If the level update isn't propagated to all loggers, then a subsequent `factory.SetLevel("x", Off)` won't silence the messages above.

I have an improved version that fixes propagation but I'm still working on the tests. Stay tuned.